extensions/ql-vscode/src/pure/files.ts

```diff
@@ -88,3 +88,26 @@ export async function readDirFullPaths(path: string): Promise<string[]> {
   const baseNames = await readdir(path);
   return baseNames.map((baseName) => join(path, baseName));
 }
+
+/**
+ * Recursively walk a directory and return the full path to all files found.
+ * Symbolic links are ignored.
+ *
+ * @param dir the directory to walk
+ *
+ * @return An iterator of the full path to all files recursively found in the directory.
+ */
+export async function* walkDirectory(
+  dir: string,
+): AsyncIterableIterator<string> {
+  const seenFiles = new Set<string>();
+  for await (const d of await opendir(dir)) {
+    const entry = join(dir, d.name);
+    seenFiles.add(entry);
+    if (d.isDirectory()) {
+      yield* walkDirectory(entry);
+    } else if (d.isFile()) {
+      yield entry;
+    }
+  }
+}
```

Comment:
It looks like only the codeql cli is using this, can we move it into the codeql-cli folder? If not: since we want to phase out the pure folder we should move it into common at least?

Reply:
I think for now it makes the most sense to put it in this file, just so it's related to everything else. I agree that we should move it into the `common` folder if we're going to phase out the `pure` folder, but I don't think splitting up the file handling functions between two files makes this easier to read and understand.

Reply:
That makes, sense, best to move the whole file to common!

Reply:
I agree, but I don't think it's part of this PR, I think deprecating/moving files out of the `pure` directory is a separate item.